// react-3d-carousel/js/Carousel.js
'use strict';
var React = require('react');

var Util = require('./util');
var Layout = require('./layout');
var Depot = require('./depot');

var Carousel = React.createClass({displayName: "Carousel",
    getInitialState: function () {
        return {
            cards: this.props.cards,
            figures: Layout[this.props.layout].figures(this.props.width, this.props.cards, 0),
            rotationY: 0
        };
    },
    componentWillMount: function () {
        this.depot = Depot(this.getInitialState(), this.props, this.setState.bind(this));
        this.onRotate = this.depot.onRotate.bind(this);
    },
    componentWillReceiveProps: function (nextProps) {
        this.depot.onNextProps(nextProps);
    },
    render: function () {
        var angle = (2 * Math.PI) / this.state.figures.length;
        var translateZ = -Layout[this.props.layout].distance(this.props.width,
            this.state.figures.length);
        var figures = this.state.figures.map(function (d, i) {
            return (React.createElement("figure", {key: i, style: Util.figureStyle(d)},
               React.createElement("div", {className: "ui card"},
                 React.createElement("div", {className: "image"},
                   React.createElement("img", {src: d.card.hash.image_url})
                 ),
                 React.createElement("div", {className: "content"},
                   React.createElement("a", {className: "header"}, d.card.hash.name),
                   React.createElement("div", {className: "meta"},
                   React.createElement("span", {className: "date"}, "Joined in 2013")
                   ),
                   React.createElement("div", {className: "description"},
                     d.card.hash.categories[0]
                   )
                 ),
                 React.createElement("div", {className: "extra content"},
                   React.createElement("a", null,
                   React.createElement("i", {className: "user icon"}),
                   d.card.hash.rating
                   )
                 )
               )
               ));
        });
        return (
          React.createElement("div", null,
            React.createElement("section", {className: "react-3d-carousel"},
            React.createElement("div", {className: "carousel",
            style: {transform: "translateZ("+translateZ+"px)"}},
            figures
            ),
            React.createElement("div", {className: "prev", onClick: Util.partial(this.onRotate,+angle)}),
            React.createElement("div", {className: "next", onClick: Util.partial(this.onRotate,-angle)})
            ),

            React.createElement("button", {className: "massive ui button", onClick: Util.partial(this.onRotate,-(19*angle))}, "Spin")
          )
            );
    }
});
module.exports = Carousel;
