// react-3d-carousel/js/Carousel.js
'use strict';
var React = require('react');

var Util = require('./util');
var Layout = require('./layout');
var Depot = require('./depot');

var Carousel = React.createClass({
    displayName: 'Carousel',

    getInitialState: function getInitialState() {
        return {
            images: this.props.images,
            figures: Layout[this.props.layout].figures(this.props.width, this.props.images, 0),
            rotationY: 0
        };
    },
    componentWillMount: function componentWillMount() {
        this.depot = Depot(this.getInitialState(), this.props, this.setState.bind(this));
        this.onRotate = this.depot.onRotate.bind(this);
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        this.depot.onNextProps(nextProps);
    },
    render: function render() {
        var angle = 2 * Math.PI / this.state.figures.length;
        var translateZ = -Layout[this.props.layout].distance(this.props.width, this.state.figures.length);
        var figures = this.state.figures.map(function (d, i) {
            return (React.createElement("figure", {key: i, style: Util.figureStyle(d)},

               React.createElement("div", {className: "ui card"},
               React.createElement("div", {className: "image"},
               React.createElement("img", {src: d.image})
               ),
               React.createElement("div", {className: "content"},
               React.createElement("a", {className: "header"}, "Kristy"),
               React.createElement("div", {className: "meta"},
               React.createElement("span", {className: "date"}, "Joined in 2013")
               ),
               React.createElement("div", {className: "description"},
               "Kristy is an art director living in New York."
               )
               ),
               React.createElement("div", {className: "extra content"},
               React.createElement("a", null,
               React.createElement("i", {className: "user icon"}),
               "22 Friends"
               )
               )
               )
               ));
        });
        return React.createElement(
            'section',
            { className: 'react-3d-carousel' },
            React.createElement(
                'div',
                { className: 'carousel',
                    style: { transform: 'translateZ(' + translateZ + 'px)' } },
                figures
            ),
            React.createElement('div', { className: 'prev', onClick: Util.partial(this.onRotate, +angle) }),
            React.createElement('div', { className: 'next', onClick: Util.partial(this.onRotate, -angle) })
        );
    }
});
module.exports = Carousel;
