// react-3d-carousel/src/Carousel.jsx
'use strict';
var React = require('react');

var Util = require('./util');
var Layout = require('./layout');
var Depot = require('./depot');

var Carousel = React.createClass({
    getInitialState: function () {
        return {
            cards: this.props.cards,
            figures: Layout[this.props.layout].figures(this.props.width, this.props.cards, 0),
            rotationY: 0
        };
    },
    componentWillMount: function () {
        this.depot = Depot(this.getInitialState(), this.props, this.setState.bind(this));
        this.onRotate = this.depot.onRotate.bind(this);
    },
    componentWillReceiveProps: function (nextProps) {
        this.depot.onNextProps(nextProps);
    },
    render: function () {
        var angle = (2 * Math.PI) / this.state.figures.length;
        var translateZ = -Layout[this.props.layout].distance(this.props.width,
            this.state.figures.length);
        var figures = this.state.figures.map(function (d, i) {
            return (<figure key={i} style={Util.figureStyle(d)}>
               <div className="ui card">
                 <div className="image">
                   <img src={d.card.image} />
                 </div>
                 <div className="content">
                   <a className="header">{d.card.name}</a>
                   <div className="meta">
                   <span className="date">Joined in 2013</span>
                   </div>
                   <div className="description">
                     {d.card.description}
                   </div>
                 </div>
                 <div className="extra content">
                   <a>
                   <i className="user icon"></i>
                   {d.card.rating}
                   </a>
                 </div>
               </div>
               </figure>);
        });
        return (
            <section className='react-3d-carousel'>
            <div className='carousel'
            style={{transform: "translateZ("+translateZ+"px)"}}>
            {figures}
            </div>
            <div className='prev' onClick={Util.partial(this.onRotate,+angle)}></div>
            <div className='next' onClick={Util.partial(this.onRotate,-angle)}></div>
            </section>
            );
    }
});
module.exports = Carousel;
