// react-3d-carousel/src/Carousel.jsx
'use strict';
var React = require('react');

var Util = require('./util');
var Layout = require('./layout');
var Depot = require('./depot');

var Carousel = React.createClass({
    getInitialState: function () {
        return {
            cards: this.props.cards,
            figures: Layout[this.props.layout].figures(this.props.width, this.props.cards, 0),
            rotationY: 0
        };
    },
    componentWillMount: function () {
        this.depot = Depot(this.getInitialState(), this.props, this.setState.bind(this));
        this.onRotate = this.depot.onRotate.bind(this);
    },
    componentWillReceiveProps: function (nextProps) {
        this.depot.onNextProps(nextProps);
    },
    render: function () {
        var angle = (2 * Math.PI) / this.state.figures.length;
        var translateZ = -Layout[this.props.layout].distance(this.props.width,
            this.state.figures.length);
        var figures = this.state.figures.map(function (d, i) {
            return (<figure key={i} style={Util.figureStyle(d)}>
               <div className="ui card">
                 <div className="image">
                   <img src={d.card.hash.image_url} />
                 </div>
                 <div className="content">
                   <a className="header" href={d.card.hash.url} target="_new">{d.card.hash.name}</a>
                   <div className="description">
                     {d.card.hash.categories[0][0]}
                   </div>
                 </div>
                 <div className="extra content">
                   <span>
                    <i className="orange star icon"></i>{d.card.hash.rating}
                   </span>
                   <span className="right floated">
                     <a href={"https://maps.google.com/maps?q="+d.card.hash.location.display_address[0] + " " + d.card.hash.location.display_address[1] } target="_blank"><i className="orange map icon"></i>{d.card.hash.location.display_address[0]}</a>
                   </span>
                 </div>
               </div>
               </figure>);
        });
        return (
          <div>
            <section className='react-3d-carousel'>
            <div className='carousel'
            style={{transform: "translateZ("+translateZ+"px)"}}>
            {figures}
            </div>
            <div className='prev' onClick={Util.partial(this.onRotate,+angle)}></div>
            <div className='next' onClick={Util.partial(this.onRotate,-angle)}></div>
            </section>
            <div className='spin-button'>
              <button className='massive ui red button' onClick={Util.partial(this.onRotate,-((Math.floor(Math.random() * (39 - 19 + 1)) + 19)*angle))}>Spin</button>
            </div>
          </div>
        );
    }
});
module.exports = Carousel;
